src/content/configuration/output.md

```diff
@@ -422,23 +422,93 @@ Note this option does not affect output files for on-demand-loaded chunks. For t
+
+| Template   | Description                                        |
+| ---------- | -------------------------------------------------- |
+| [file]     | Filename and path, without query or fragment       |
+| [query]    | Query with leading `?`                             |
+| [fragment] | Fragment with leading `#`                          |
+| [base]     | Only filename (including extensions), without path |
+| [filebase] | Same, but deprecated                               |
+| [path]     | Only path, without filename                        |
+| [name]     | Only filename without extension or path            |
+| [ext]      | Extension with leading `.`                         |
+
+Substituions available on URL-level:
+
+| Template | Description |
+| -------- | ----------- |
+| [url]    | URL         |
+
+T> `[file]` equals `[path][base]`. `[base]` equals `[name][ext]`. The full path is `[path][name][ext][query][fragment]` or `[path][base][query][fragment]` or `[file][query][fragment]`.
+
+The lengths of hashes (`[hash]`, `[contenthash]` or `[chunkhash]`) can be specified using `[hash:16]` (defaults to 20). Alternatively, specify [`output.hashDigestLength`](#outputhashdigestlength) to configure the length globally.
 
 It is possible to filter out placeholder replacement when you want to use one of the placeholders in the actual file name. For example, to output a file `[name].js`, you have to escape the `[name]` placeholder by adding backslashes between the brackets. So that `[\name\]` generates `[name]` instead of getting replaced with the `name` of the asset.
 
 Example: `[\id\]` generates `[id]` instead of getting replaced with the `id`.
 
-If using a function for this option, the function will be passed an object containing the substitutions in the table above.
+If using a function for this option, the function will be passed an object containing data for the substitutions in the table above.
+Substituions will be applied to the returned string too.
+The passed object will have this type: (properties available depending on context)
+
+``` typescript
+type PathData = {
+  hash: string,
+  hashWithLength: (number) => string,
+  chunk: Chunk | ChunkPathData,
+  module: Module | ModulePathData,
+  contentHashType: string,
+  contentHash: string,
+  contentHashWithLength: (number) => string,
+  filename: string,
+  url: string,
+  runtime: string | SortableSet<string>,
+  chunkGraph: ChunkGraph
+}
+type ChunkPathData = {
+  id: string | number,
+  name: string,
+  hash: string,
+  hashWithLength: (number) => string,
+  contentHash: Record<string, string>,
+  contentHashWithLength: Record<string, (number) => string>
+}
+type ModulePathData = {
+  id: string | number,
+  hash: string,
+  hashWithLength: (number) => string
+}
+```
+
+T> In some context properties will use javascript code expressions instead of raw values. In these cases the `WithLength` variant is available and should be used instead of slicing the original value.
```

Comment:
```suggestion
T> In some context properties will use JavaScript code expressions instead of raw values. In these cases the `WithLength` variant is available and should be used instead of slicing the original value.
```